doc/designs/thread-api.md

```diff
@@ -0,0 +1,305 @@
+Thread Pool Support
+===================
+
+OpenSSL wishes to support the internal use of threads for purposes of
+concurrency and parallelism in some circumstances. There are various reasons why
+this is desirable:
+
+  - Some algorithms are designed to be run in parallel (Argon2);
+  - Some transports (e.g. QUIC, DTLS) may need to handle timer events
+    independently of application calls to OpenSSL.
+
+To support this end, OpenSSL can manage an internal thread pool. Tasks can be
+scheduled on the internal thread pool.
+
+There are three options given to an application which wants to use the thread
+pool functionality:
+
+  1. Use OpenSSL's default thread pool functionality.
+
+  2. Use OpenSSL's thread pool functionality with a custom thread spawner
+     method.
+
+  3. Provide threads manually to OpenSSL for use as part of its thread pool.
+
+Default Model
+-------------
+
+In the default model, OpenSSL creates and manages threads up to a maximum
+number of threads authorized by the application.
+
+The application enables thread pooling by calling the following function
+during its initialisation:
+
+```c
+/*
+ * Set the maximum number of threads to be used by the thread pool.
+ *
+ * If the argument is 0, thread pooling is disabled. OpenSSL will not create any
+ * threads and existing threads in the thread pool will be torn down.
+ *
+ * Returns 1 on success and 0 on failure. Returns failure if OpenSSL-managed
+ * thread pooling is not supported (for example, if it is not supported on the
+ * current platform, or because OpenSSL is not built with the necessary
+ * support).
+ */
+int OSSL_set_max_threads(uint64_t max_threads);
+
+/*
+ * Get the maximum number of threads currently allowed to be used by the
+ * thread pool. If thread pooling is disabled or not available, returns 0.
+ */
+uint64_t OSSL_get_max_threads(void);
+```
+
+The maximum thread count is a limit, not a target. Threads will not be spawned
+unless (and until) there is demand.
```

Comment:
What happens in the case that `OSSL_set_max_threads` is used to lower an existing limit? Presumably existing threads will finish their current work item?

What happens if a set of threads are idle for a while? Do we stop them or keep them hanging around just in case we might need them?

Since start-up of threads is (presumably) expensive an application might want to have the  control over when the threads in the thread pool are created - rather than leaving it  them being created on demand.

I could envisage different flags that might be set for different thread management policies. e.g. a flag to start threads "eagerly", a flag to keep threads around even if they are idle, etc. There might be some other values involved too, e.g. a thread timeout value (how long to wait before stopping idle threads)

Reply:
>Presumably existing threads will finish their current work item?

That was what I had in mind, yes.

>What happens if a set of threads are idle for a while? Do we stop them or keep them hanging around just in case we might need them?

I believe this would not be defined by the API contract and it would be for us to decide what strategy to use here, which could change between releases. Since OpenSSL itself presumably wants to avoid excessive thread creation cost I imagine it would have a heuristic for when to tear down threads which aren't being used which performs well.

If an application wants to control thread creation, this could obviously be covered by the manual API above.

For things like idle thread timeout which serve as inputs to OpenSSL's internal thread management strategy, this sounds like a good idea, but being essentially a tuning knob, should probably wait for beyond MVP.

Being able to force creation or maintenance of idle threads etc. is a good idea and worth considering, though again there's a question here of how tightly we want to constrain our scope for an MVP threading proposal.

Reply:
Shouldn't these functions accept OSSL_LIB_CTX * as well?